src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -96,5 +95,18 @@ void AddDocumentsFromProject(Project? project, bool isOpen)
                 result.AddRange(project.Documents);
             }
         }
+
+        protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Document document, Option2<DiagnosticMode> diagnosticMode, CancellationToken cancellationToken)
+        {
+            // We only support workspace diagnostics for closed files.
+            if (document.IsOpen())
+                return SpecializedTasks.EmptyImmutableArray<DiagnosticData>();
+
+            // For closed files, go to the IDiagnosticService for results.  These won't necessarily be totally up to
+            // date.  However, that's fine as these are closed files and won't be in the process of being edited.  So
+            // any deviations in the spans of diagnostics shouldn't be impactful for the user.
+            var diagnostics = this.DiagnosticService.GetPullDiagnostics(document, includeSuppressedDiagnostics: false, diagnosticMode, cancellationToken);
```

Comment:
this split actually makes a fair amount of sense to me then.  Workspace diagnostics use whatever is in the workspace (which the server owns as these are closed files) whereas files with content managed via LSP (open) force a calculation based on the client's known content.

The only issue I can see is if the content in the open file changes the workspace diagnostics (e.g. delete a method in an open file that's used elsewhere).  I guess in that case the experience would that the error list could show stale errors until the next workspace/diagnostic message comes in that acts on the new workspace content.

But squiggles are always fine as they act on open documents (and therefore LSP content)